Issue #834: Server crash in print_keydup_error / key_unpack or unexpe…

spetrunia opened the pull request:
…cted ER_DUP_KEY

Fix two issues:
1. Rdb_ddl_manager::rename() loses the value of m_hidden_pk_val. new
object used to get 0, which means "not loaded from the db yet".

2. ha_rocksdb::load_hidden_pk_value() uses current transaction (and its
snapshot) when loading hidden PK value from disk. This may cause it to
load an out-of-date value.

laurynas-biveinis commented on mysql-test/suite/rocksdb/t/allow_no_primary_key.test:
```diff
@@ -96,3 +96,31 @@ DELETE FROM t1 WHERE a = 35 AND b = 'foo';
+--echo #
+CREATE TABLE t1 (a INT, KEY(a)) ENGINE=RocksDB;
+INSERT INTO t1 VALUES (1),(1+1);
+create table t2 (a int);
```
Do I understand correctly that t2 engine is not important as long as it's transactional?

spetrunia replied:
No, I think it needs to be MyRocks.
(t2 is present in the test to cause MyRocks to start a transaction. It is a MyRocks table in my setup. I didn't try other engines for it but I don't think using another engine will cause MyRocks transaction to be created)

I'll add ENGINE=RocksDB.

hermanlee commented on mysql-test/suite/rocksdb/r/allow_no_primary_key.result:
```diff
@@ -262,3 +262,34 @@ SELECT * FROM t1;
+10
+20
+30
+connect  con1,localhost,root,,;
```
Could you re-record the test? I do not see these lines in the test output and the test itself fails.

spetrunia replied:
Yes, will do.

hermanlee commented on storage/rocksdb/ha_rocksdb.cc:
```diff
@@ -5216,8 +5216,17 @@ int ha_rocksdb::load_hidden_pk_value() {
-  Rdb_transaction *const tx = get_or_create_tx(table->in_use);
-  const bool is_new_snapshot = !tx->has_snapshot();
+  /*
+    We should read the latest committed value in the database.
```
Is this needed, or is this being extra defensive, but not strictly required? It seems the fix needed is transferring m_hidden_pk_val from the old table_def to the new table_def during alter, and that is already in this diff. This code only populates the m_hidden_pk_val of the table if it hasn't been loaded yet. If there is an open transaction with the table, then we shouldn't be able to rename the table. If the current snapshot does not read the latest data, then another transaction must have written the new m_hidden_pk_val after this snapshot was created. In that case, the table definition should have the latest data, and whether we read with the latest snapshot or the existing snapshot shouldn't matter, no?

spetrunia replied:
Also discovered this with valgrind: if one calls `tx->start_tx`, they must call set_params beforehand, like it's done at other places:
```
    tx->set_params(THDVAR(thd, lock_wait_timeout), THDVAR(thd, max_row_locks));
    tx->start_tx();
```

spetrunia replied:
... as discussed on the call - the patch is extra defensive, and I will remove this part.  Will leave only the one in `Rdb_ddl_manager::rename`.

spetrunia committed:
Issue #834: Server crash in print_keydup_error / key_unpack or unexpected ER_DUP_KEY

The issue is caused by a combination of two factors:

1. Rdb_ddl_manager::rename() loses the value of m_hidden_pk_val. new
object used to get 0, which means "not loaded from the db yet".

2. ha_rocksdb::load_hidden_pk_value() uses current transaction (and its
snapshot) when loading hidden PK value from disk. This may cause it to
load an out-of-date value.

This patch fixes #1, which is sufficient to fix the issue.
(This is attempt 2, with review input addressed)
